accepted/single-file/design.md

```diff
@@ -0,0 +1,279 @@
+
+## Design
+
+There are two main aspects to publishing apps as a self-extracting single file:
+
+- The Bundler: A tool that embeds the managed app, its dependencies, and the runtime into a single host executable.
+- The host: The "single-file" which facilitates the extraction and/or loading of embedded components.
+
+### The Bundler
+
+#### Bundling Tool
+
+The bundler tool must be:
+
+* Able to embed any file required to run a .Net Core app -- managed assemblies, native binaries, configuration files, data files, etc.
+* Able to generate cross-platform bundles (ex: publish a single-file for Linux target from Windows)
+* Deterministic (generate the exact same single-file on multiple runs)
+
+With the above requirements in mind, we plan to implement a tool similar to [MkBundle](https://github.com/mono/mono/blob/master/mcs/tools/mkbundle) that simply appends the bundled dependencies as a binary blob at the end of a host binary.
+
+To any host (native binary) specified, the bundler tool will add:
+
+* A footer that contains:
+  * A flag to identify that this is actually a bundle.
+  * A version identifier.
+  * Offset of the bundle manifest
+* A manifest that identifies (via offset and size)
+  * The configuration files `app.deps.json` `app.runtimeconfig.json`
+  * The embedded app to load
+  * The set of MSIL images
+  * The set of ready-to-run images
+  * All other files
+* The actual files to be bundled.
+
+##### Implementation
+
+The bundler should ideally be located close to the dotnet host, since their implementation is closely related. Therefore, the bundler will be implemented in the `core-setup` repo. The `corehost` code is expected to move to the CoreCLR repo, at which point, the bundler should move with it. The bundler will be implemented in managed code.
+
+##### Command Line Interface
+
+The bundler will be a standalone tool with the following command line:
+
+```bash
+bundle -a <App>      The name of the managed app
+       -h <Host>     The path to the native host
+       -c <contents> The path to a directory containing the files to bundle
+       [-o <output>] The path to output bundle
+       [-v]          Generate verbose output
+```
+
+Most users are only expected to interact with the bundler via the build system.
+
+##### Build System Interface
+
+Publishing to a single file can be triggered by adding the following property to an application's project file:
+
+```xml
+<PropertyGroup>
+    <PublishSingleFile>true</PublishSingleFile>
+</PropertyGroup>    
+```
+
+* The `PublishSingleFile` property applies to both framework dependent and self-contained publish operations.
+* Setting the `PublishSingleFile`property causes the managed app, dependencies, configurations, etc. (basically the contents of the publish directory when `dotnet publish` is run without setting the property) to be embedded within the native `apphost`. The publish directory will only contain the single bundled executable (and the symbol file).
+
```

Comment:
That is actually a concern for me. I want to be able to have a single package that include the PDB. We currently do that with embedded pdbs, and I assume this will continue to work?

Reply:
Thanks for the feedback, we could make this configurable.  I think by default (to save size on disk) this would be off.

Reply:
@ayende you're asking about assemblies generated with debug info embedded  (ex: `csc -debug:embedded`) right? Those are fine, I'll add it as a scenario to test.
 If a separate PDB file is generated, we won't add it to the single-file. But as @jeffschwMSFT says, we can make it configurable if necessary.

Reply:
+1 on configurably embedded PDB.

Reply:
Yes, I am. If those work, I'm happy with that.